### DISCUSSION TITLE
Trying to save output file but it saves as directory , not file

### DISCUSSION THREAD

**Discussion by Stevew7777** [ultralytics/yolov5#12354] [General]
Hi

I'm trying to get YOLOv5 to save its processed output as a single file in a specific location that I want 
e.g. /tmp/output/myoutput.jpg  , but instead it saves it as a directory called /tmp/output/myoutput.jpg/ with an output file called myoutput.jpg   

Not sure what I'm doing wrong? 

Running YOLOv5    I use this command :

python /usr/src/app/detect.py --weights /usr/src/app/runs/train/exp7/weights/last.pt --source /tmp/input/input.jpg   --name /tmp/output/myoutput.jpg   --exist-ok

The config inside detect.py is :

def run(
        weights=ROOT / 'yolov5s.pt',  # model path or triton URL
        source=ROOT / 'data/images',  # file/dir/URL/glob/screen/0(webcam)
        data=ROOT / 'data/coco128.yaml',  # dataset.yaml path
        imgsz=(640, 640),  # inference size (height, width)
        conf_thres=0.25,  # confidence threshold
        iou_thres=0.45,  # NMS IOU threshold
        max_det=1000,  # maximum detections per image
        device='',  # cuda device, i.e. 0 or 0,1,2,3 or cpu
        view_img=False,  # show results
        save_txt=False,  # save results to *.txt
        save_conf=False,  # save confidences in --save-txt labels
        save_crop=False,  # save cropped prediction boxes
        nosave=False,  # do not save images/videos
        classes=None,  # filter by class: --class 0, or --class 0 2 3
        agnostic_nms=False,  # class-agnostic NMS
        augment=False,  # augmented inference
        visualize=False,  # visualize features
        update=False,  # update all models
        project=ROOT / '/tmp/output/',  # save results to project/name
        name='',  # save results to project/name
        exist_ok=False,  # existing project/name ok, do not increment
        line_thickness=3,  # bounding box thickness (pixels)
        hide_labels=False,  # hide labels
        hide_conf=False,  # hide confidences
        half=False,  # use FP16 half-precision inference
        dnn=False,  # use OpenCV DNN for ONNX inference
        vid_stride=1,  # video frame-rate stride

def parse_opt():
    parser = argparse.ArgumentParser()
    parser.add_argument('--weights', nargs='+', type=str, default=ROOT / 'yolov5s.pt', help='model path or triton URL')
    parser.add_argument('--source', type=str, default=ROOT / 'data/images', help='file/dir/URL/glob/screen/0(webcam)')
    parser.add_argument('--data', type=str, default=ROOT / 'data/coco128.yaml', help='(optional) dataset.yaml path')
    parser.add_argument('--imgsz', '--img', '--img-size', nargs='+', type=int, default=[640], help='inference size h,w')
    parser.add_argument('--conf-thres', type=float, default=0.25, help='confidence threshold')
    parser.add_argument('--iou-thres', type=float, default=0.45, help='NMS IoU threshold')
    parser.add_argument('--max-det', type=int, default=1000, help='maximum detections per image')
    parser.add_argument('--device', default='', help='cuda device, i.e. 0 or 0,1,2,3 or cpu')
    parser.add_argument('--view-img', action='store_true', help='show results')
    parser.add_argument('--save-txt', action='store_true', help='save results to *.txt')
    parser.add_argument('--save-conf', action='store_true', help='save confidences in --save-txt labels')
    parser.add_argument('--save-crop', action='store_true', help='save cropped prediction boxes')
    parser.add_argument('--nosave', action='store_true', help='do not save images/videos')
    parser.add_argument('--classes', nargs='+', type=int, help='filter by class: --classes 0, or --classes 0 2 3')
    parser.add_argument('--agnostic-nms', action='store_true', help='class-agnostic NMS')
    parser.add_argument('--augment', action='store_true', help='augmented inference')
    parser.add_argument('--visualize', action='store_true', help='visualize features')
    parser.add_argument('--update', action='store_true', help='update all models')
    parser.add_argument('--project', default=ROOT / '/tmp/docker_upload', help='save results to project/name')
    parser.add_argument('--name', default='', help='save results to project/name')
    parser.add_argument('--exist-ok', action='store_true', help='existing project/name ok, do not increment')
    parser.add_argument('--line-thickness', default=3, type=int, help='bounding box thickness (pixels)')
    parser.add_argument('--hide-labels', default=False, action='store_true', help='hide labels')
    parser.add_argument('--hide-conf', default=False, action='store_true', help='hide confidences')
    parser.add_argument('--half', action='store_true', help='use FP16 half-precision inference')
    parser.add_argument('--dnn', action='store_true', help='use OpenCV DNN for ONNX inference')
    parser.add_argument('--vid-stride', type=int, default=1, help='video frame-rate stride')
    opt = parser.parse_args()
    opt.imgsz *= 2 if len(opt.imgsz) == 1 else 1  # expand
    print_args(vars(opt))
    return opt

Any help appreciated.  :-)

**Comment by Stevew7777** [comment]
Anyone? This cant be too hard?

**Comment by OJJJN** [comment]
maybe try this, you need to improve some code:

import cv2
import numpy as np
import argparse
import time
import datetime
import os

model = 'best.onnx'
img_w = 640
img_h = 640
classes_file = 'classes.txt'

def class_names():
    classes = []
    with open(classes_file, 'r') as file:
        for line in file:
            name = line.strip('\n')
            classes.append(name)
    return classes

width_frame = 640
net = cv2.dnn.readNetFromONNX(model)
classes = class_names()

# Dahua CCTV camera settings
rtsp_url = ''  # Replace with your Dahua CCTV camera RTSP URL

cap = cv2.VideoCapture(rtsp_url)

total_slots = 15
cars_count = 0
empty_slots = 0
last_detection_time = datetime.datetime.now()

# Specify the directory path to save the image results
output_directory = r'D:\Database'

while True:
    current_time = datetime.datetime.now()
    time_difference = current_time - last_detection_time

    if time_difference.total_seconds() >= (15):
        detection_start_time = datetime.datetime.now()
        detection_end_time = detection_start_time + datetime.timedelta(seconds=5)
        last_detection_time = detection_end_time

        while datetime.datetime.now() <= detection_end_time:
            ret, frame = cap.read()

            if not ret:
                print("Failed to capture frame")
                break

            height = int(frame.shape[0] * (width_frame / frame.shape[1]))
            dim = (width_frame, height)
            img = cv2.resize(frame, dim, interpolation=cv2.INTER_AREA)

            blob = cv2.dnn.blobFromImage(img, 1/255, (img_w, img_h), swapRB=True, mean=(0, 0, 0), crop=False)
            net.setInput(blob)
            t1 = time.time()
            outputs = net.forward(net.getUnconnectedOutLayersNames())
            t2 = time.time()
            out = outputs[0]
            n_detections = out.shape[1]
            height, width = img.shape[:2]
            x_scale = width / img_w
            y_scale = height / img_h
            conf_threshold = 0.7
            score_threshold = 0.5
            nms_threshold = 0.5

            class_ids = []
            score = []
            boxes = []

            for i in range(n_detections):
                detect = out[0][i]
                confidence = detect[4]
                if confidence >= conf_threshold:
                    class_score = detect[5:]
                    class_id = np.argmax(class_score)
                    if class_score[class_id] > score_threshold:
                        score.append(confidence)
                        class_ids.append(class_id)
                        x, y, w, h = detect[0], detect[1], detect[2], detect[3]
                        left = int((x - w/2) * x_scale)
                        top = int((y - h/2) * y_scale)
                        width = int(w * x_scale)
                        height = int(h * y_scale)
                        box = np.array([left, top, width, height])
                        boxes.append(box)

            indices = cv2.dnn.NMSBoxes(boxes, np.array(score), conf_threshold, nms_threshold)

            cars_count = len(indices)
            empty_slots = total_slots - cars_count

            for i in indices:
                box = boxes[i]
                left = box[0]
                top = box[1]
                width = box[2]
                height = box[3]

                left, top, width, height = boxes[i]

                cv2.rectangle(img, (left, top), (left + width, top + height), (0, 0, 255), 2)
                label = "{}".format(classes[class_ids[i]])
                text_size = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
                dim, baseline = text_size[0], text_size[1]
                cv2.rectangle(img, (left, top - 20), (left + dim[0], top + dim[1] + baseline - 20), (0, 0, 0), cv2.FILLED)
                cv2.putText(img, label, (left, top + dim[1] - 20), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 0), 1, cv2.LINE_AA)

            # Display car count and empty slots on theframe
            cv2.putText(img, f"Cars Count: {cars_count}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (255, 255, 255), 2)[0]
            cv2.putText(img, f"Empty Slots: {empty_slots}", (10, 70), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (255, 255, 255), 2)[0]

            # Save the image results
            filename = f"result_{datetime.datetime.now().strftime('%Y%m%d_%H%M%S')}.jpg"
            output_path = os.path.join(output_directory, filename)
            cv2.imwrite(output_path, img)

            # Display the captured frame
            cv2.imshow("Object Detection", img)

            # Check for 'q' key press to exit
            if cv2.waitKey(1) & 0xFF == ord('q'):
                break

    else:
        ret, frame = cap.read()

        if not ret:
            print("Failed to capture frame")
            break

        # Display the captured frame
        cv2.imshow("Camera Feed", frame)

        # Check for 'q' key press to exit
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

# Release the video capture and destroy all windows
cap.release()
cv2.destroyAllWindows()

this is using opencv and onnx file, you can try to change the onnx required or something. also this code saving an image with date and time name, so you know when the object got detect.

**Reply by OJJJN** [reply]
this code will run every 15 sec, so it not consume so many cpu or gpu
